### PR TITLE
feat: implement the X Input extension to support touch input, smooth scrolling, etc.

### LINKED ISSUE #17912: Scroll using touch screen isn't working
### Current behavior

Any scrollable section (ListView, ScrollViewer) can't be scrolled on Skia/X11 (Desktop) using gestures (touch screen). Worked fine in 5.1 and lower. Tested SDK 5.3.96 and 5.4.0-dev.180, same issue.

### Expected behavior

Dragging a scrollable section using finger should scroll the content

### How to reproduce it (as minimally and precisely as possible)

Create new Uno App, in `MainPage.xaml` replace the TextBox and Button with this content:
```
        <TextBox Text="{Binding Name, Mode=TwoWay}"
            PlaceholderText="Enter your name:" />
        <ScrollViewer HorizontalScrollMode="Disabled" VerticalScrollBarVisibility="Visible" Height="100">
          <StackPanel Orientation="Vertical">
            <Button Margin="8">Hej 1</Button>
            <Button Margin="8">Hej 2</Button>
            <Button Margin="8">Hej 3</Button>
            <Button Margin="8">Hej 4</Button>
            <Button Margin="8">Hej 5</Button>
          </StackPanel>
        </ScrollViewer>
        <Button Content="Go to Second Page"
            AutomationProperties.AutomationId="SecondPageButton"
            Command="{Binding GoToSecond}" />
```
And run on Desktop/Skia and try to scroll the section using touch screen. Note that I have only tested this on Raspberry Pi with a touch screen, I don't have touch on my Windows laptop.

### Workaround

_No response_

### Works on UWP/WinUI

None

### Environment

Uno.UI / Uno.UI.WebAssembly / Uno.UI.Skia

### NuGet package version(s)

SDK 5.3.96
SDK 5.4.0-dev.180

### Affected platforms

Skia (Linux X11)

### IDE

Visual Studio 2022

### IDE version

_No response_

### Relevant plugins

_No response_

### Anything else we need to know?

_No response_

### PR DESCRIPTION
**Pull request by ramezgerges**
GitHub Issue (If applicable): closes https://github.com/unoplatform/uno-private/issues/409, closes #17912, closes https://github.com/unoplatform/uno/issues/16804

## PR Type

What kind of change does this PR introduce?

## PR Checklist

Please check if your PR fulfills the following requirements:

- [ ] Docs have been added/updated which fit [documentation template](https://github.com/unoplatform/uno/blob/master/doc/.feature-template.md) (for bug fixes / features)
- [ ] [Unit Tests and/or UI Tests](https://github.com/unoplatform/uno/blob/master/doc/articles/uno-development/working-with-the-samples-apps.md) for the changes have been added (for bug fixes / features) (if applicable)
- [ ] Validated PR `Screenshots Compare Test Run` results.
- [ ] Contains **NO** breaking changes
- [ ] Associated with an issue (GitHub or internal) and uses the [automatic close keywords](https://help.github.com/en/github/managing-your-work-on-github/linking-a-pull-request-to-an-issue).
- [ ] Commits must be following the [Conventional Commits](https://www.conventionalcommits.org/en/v1.0.0/#summary) specification.

## Other information

Internal Issue (If applicable):

### CODE REVIEW

**Comment by jeromelaban** [comment]
@ramezgerges other than the rebase, is there anything missing in this PR? Once it is, I'll test it on actual hardware.

**Comment by ramezgerges** [comment]
@jeromelaban There's nothing that I plan to add, no. But considering I haven't tested touch at all, the PR will probably have some stuff "missing.

**Comment by jeromelaban** [comment]
@ramezgerges could you resolve the conflicts? Thanks!

File: src/Uno.UI.Runtime.Skia.X11/X11PointerInputSource.XInput.cs [hunk 1]
```diff
@@ -0,0 +1,513 @@
...
+			IsXButton1Pressed = (mask & (1 << XButton1)) != 0,
+			IsXButton2Pressed = (mask & (1 << XButton2)) != 0,
+			IsHorizontalMouseWheel = isHorizontalMouseWheel,
+			IsInRange = true,
+			IsTouchPad = info is { } && IsTouchpad(info.Value.Properties),
+			MouseWheelDelta = (int)Math.Round(wheelDelta)
+			// TODO: PointerUpdateKind is currently not used for anything, so we can leave it blank
+			// for now, but ideally we should probably be diffing the previous and new button states
+			// and computing an update kind from that.
+		};
```
Comment (lindexi) [inline src/Uno.UI.Runtime.Skia.X11/X11PointerInputSource.XInput.cs:179]: How about add the touch Pressure and touch width and height here? Thank you.

File: src/Uno.UI.Runtime.Skia.X11/X11PointerInputSource.XInput.cs [hunk 2]
```diff
@@ -0,0 +1,513 @@
...
+					var args = CreatePointerEventArgsFromEnterLeaveEvent(
+						ev.GenericEventCookie.GetEvent<XIEnterLeaveEvent>(),
+						PointerDeviceType.Mouse);
+					if (evtype is XiEventType.XI_Enter)
+					{
+						X11XamlRootHost.QueueAction(_host, () => RaisePointerEntered(args));
+					}
+					else
+					{
+						X11XamlRootHost.QueueAction(_host, () => RaisePointerExited(args));
```
Comment (lindexi) [inline src/Uno.UI.Runtime.Skia.X11/X11PointerInputSource.XInput.cs:266]: Should it check `XiEventType.detail` to `XiEnterLeaveDetail.XINotifyAncestor` here? The XINotifyAncestor means the application receive the leave from the ancestor window, such as dialog.

**Comment by lindexi** [comment]
My version of X input support: https://github.com/lindexi/uno/blob/f2f4d7980fcbf3cef4cbd4a56172e0cb1121aa74/src/Uno.UI.Runtime.Skia.X11/X11PointerInputSource.XInput.cs

I tested the code on a touch device and it passed.

But there are some points worth discussing:

1. Within XInput, the size of touch is no longer represented by width and height, but by the long and short axes of an ellipse.
2. Is there a need to perform DPI scaling calculations for width and height?
3. In XInput, if there is no change in the size of touch and pressure sensitivity, no values will be received. It may be necessary to use the data from the previous touch event at the lower level. Currently, my approach is to return a default value of 0 when no values are received.

**Comment by lindexi** [comment]
Update:

I reimplemente the `X11PointerInputSource.XInput.cs` to make it to support GetIntermediatePoints, which can alleviates the problem of input message stacking.

See https://github.com/lindexi/uno/blob/218e2016093af8523a1ad2b54cd169d20da09722/src/Uno.UI.Runtime.Skia.X11/X11PointerInputSource.XInput.cs#L277

Reference: https://github.com/unoplatform/uno/issues/17527

**Comment by lindexi** [comment]
@jeromelaban Sorry to be pushing, can I know any update here?

**Comment by HakanL** [comment]
@ramezgerges I've never used custom-built/debug Uno SDK before, but if you want me to try your changes please let me know and I'll figure it out.

**Comment by HakanL** [comment]
@lindexi I see, so I have to go through all the non-sample artifacts and download locally from DevOps, extract and update my project? Or am I misunderstanding the process? Sorry if this isn't the place to discuss that.

**Comment by lindexi** [comment]
@HakanL I will pull the code and build my version, see https://platform.uno/docs/articles/uno-development/building-uno-ui.html

**Comment by HakanL** [comment]
Any luck on testing this?

**Comment by MartinZikmund** [comment]
@HakanL unfortunately not working properly yet, the input gets duplicated with both mouse and touch coming in, we need to investigate further, but very soon

**Comment by lindexi** [comment]
> @HakanL unfortunately not working properly yet, the input gets duplicated with both mouse and touch coming in, we need to investigate further, but very soon

@MartinZikmund Do you means that one touch will receive both the XI_ButtonXxx and XI_TouchXxx event? If so, you can check the `XIDeviceEvent`'s flags to know the XI_ButtonXxx from Emulated:

```
if ((xiDeviceEvent->flags & XiDeviceEventFlags.XIPointerEmulated) ==
    XiDeviceEventFlags.XIPointerEmulated)
{
    ... // Ignore
}
```

**Review by MartinZikmund — APPROVED** [review]

File: src/Uno.UI.Runtime.Skia.X11/X11PointerInputSource.XInput.cs (hunk 1 above)
Comment (ramezgerges) [reply src/Uno.UI.Runtime.Skia.X11/X11PointerInputSource.XInput.cs]: We can do MT protocol (https://www.kernel.org/doc/html/latest/input/multi-touch-protocol.html)-related stuff in a separate PR if there's enough demand.

File: src/Uno.UI.Runtime.Skia.X11/X11PointerInputSource.XInput.cs (hunk 2 above)
Comment (ramezgerges) [reply src/Uno.UI.Runtime.Skia.X11/X11PointerInputSource.XInput.cs:393]: We currently never use an Uno window as a modal/dialog window. But noted for future reference. Thanks!

**Comment by ramezgerges** [comment]
Thanks @lindexi for your insights with this PR.

**Comment by HakanL** [comment]
Excited to see this merged in! Can you comment on which Nuget pre-release package versions I can find this in to do some tests?

**Comment by ramezgerges** [comment]
@HakanL It hasn't been automatically published yet, but it should be soon. The current latest sdk version on nuget.org is 5.6.0-dev.10, so most likely the first version after that.

**Comment by HakanL** [comment]
@ramezgerges Can we expect this to be back-ported to 5.5?

**Comment by jeromelaban** [comment]
> @ramezgerges Can we expect this to be back-ported to 5.5?

@HakanL It will go into the next stable release. You can use the prerelease bits in the meantime.

**Comment by HakanL** [comment]
Yeah, it works well in 5.6 prerelease, it would just be nice to be on a stable version of Uno for my product. But it's okay, I'll wait for 5.6

**Pull request merged**